keps/sig-api-machinery/20190603-immutable-fields.md

```diff
@@ -0,0 +1,464 @@
+has to rely on either built-in validation for core types, or have to build a
+validating webhooks for CRDs.
+
+Providing a new `// +immutable` marker would help 
+- to make the API more descriptive to users
+- to help API developers by verifying these assertions automatically
+- and to publish this information via OpenAPI.
+
+## Motivation
+
+There are resources in Kubernetes which have immutable fields by design,
+i.e. after creation of an object, those fields cannot be mutated anymore. E.g. a
+pod's specification is mostly unchangeable once it is created. To change the
+pod, it must be deleted, recreated and rescheduled. Users want to implement the
+same kind of read-only semantics for CustomResources, for example:
+https://github.com/kubernetes/kubernetes/issues/65973. Today this is only possible
+with the unreasonable development overhead of a webhook.
+
+### Goals
+
+- extend the CRD API to be able to specify immutability for fields.
+- publish the immutability field of CRDs via OpenAPI as vendor extension.
+- verify immutability on CR update and patch requests.
+- propose a source code marker to be consumed by kubebuilder and openapi-gen.
+- the semantics of immutability must be driven by:
+  - we do not break/change old CRD persistence semantics.
+  - the user-observed equality used for immutability checks must match the equality on
+    persisted objects. I.e. if `StorageRoundtrip(object)` is the object returned by a
+    create or update call, then we want that `StorageRoundtrip(a) == StorageRoundtrip(b)`
+    is the equality used for comparing `a` and `b` **only modulo the order in `x-kubernetes-list-type: set` arrays**. If that check fails, a request
+    is rejected because of immutability conflict.
+- the mechanism must extend to
+  - the addition of protobuf or other encodings which unify values like empty, null and undefined.
+  - the use for existing native types in order to replace complex validation code with a simple declarative marker on the types.
+  - the restriction of the equality to only map keys, but not their values.
+  - the allowance of addition and/or deletion of map keys.
+  - the allowance of addition and/or deletion of keys in array of list-type `map`.  
+
+### Non-Goals
+
+- The mechanism must be extensible to native types, but its implementation is optional.
+- The mechanism must be extensible to future normalization behaviours which will be
+  required to support protobuf for CRs. But this KEP does not aim at defining these
+  and hence defining a custom equality which is compatible with normalization.
+- The mechanism is not supposed to allow different orders in lists to be considered equal.
+
+## Proposal
+
+We propose
+
+1. adding boolean vendor extensions to CRD OpenAPI schemas named `x-kubernetes-immutable` and `x-kubernetes-immutable-keys` with `true` as the only valid value.
+2. do **strict deep-equal** comparison of those fields marked as immutable during
```

Comment:
Now that I have read the bottom, I suggest you revise this to consider unset, empty, and 0-length items to be the same.